### PR TITLE
CLOUDP-178751 - DatabaseUser deletion protection

### PR DESCRIPTION
**Pull request by helderjs**
… controller

### All Submissions:

* [x] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?
* [ ] Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if there is one).
* [ ] Update docs/release-notes/release-notes.md if your changes should be included in the release notes for the next release.

### CODE REVIEW

File: cmd/manager/main.go [hunk 1]
```diff
@@ -303,3 +314,46 @@ func initCustomZapLogger(level, encoding string) (*zap.Logger, error) {
...
+			objectDeletionSet = true
+		}
+
+		if f.Name == "subobject-deletion-protection" {
+			subObjectDeletionSet = true
+		}
+	})
+
+	if !objectDeletionSet {
+		objDeletion := os.Getenv("OBJECT_DELETION_PROTECTION")
```
Comment (josvazg) [inline cmd/manager/main.go:337]: Nit
Just curious: should we be case insensitive here?
```suggestion
		objDeletion := strings.ToLower(os.Getenv("OBJECT_DELETION_PROTECTION"))
```

File: pkg/controller/atlasdatabaseuser/atlasdatabaseuser_controller.go [hunk 2]
```diff
@@ -78,155 +76,207 @@ func (r *AtlasDatabaseUserReconciler) Reconcile(context context.Context, req ctr
...
 	}
 
-	ctx.SetConditionTrue(status.DatabaseUserReadyType)
-	ctx.SetConditionTrue(status.ReadyType)
-	return result.ReconcileResult(), nil
-}
+	if !owner {
+		result = workflow.Terminate(
+			workflow.AtlasDeletionProtection,
+			"unable to reconcile database user: it already exists in Atlas, it was not previously managed by the operator, de deletion protection is enabled.",
```
Comment (josvazg) [inline pkg/controller/atlasdatabaseuser/atlasdatabaseuser_controller.go:144]: Typo?
```suggestion
			"unable to reconcile database user: it already exists in Atlas, it was not previously managed by the operator, and the deletion protection is enabled.",
```

File: cmd/manager/main.go [hunk 3]
```diff
@@ -303,3 +314,46 @@ func initCustomZapLogger(level, encoding string) (*zap.Logger, error) {
...
+func configureDeletionProtectionFlags(config *Config) {
+	if config == nil {
+		return
+	}
+
+	objectDeletionSet := false
+	subObjectDeletionSet := false
+
+	flag.Visit(func(f *flag.Flag) {
+		if f.Name == "object-deletion-protection" {
```
Comment (igor-karpukhin) [inline cmd/manager/main.go:327]: Consider making names of the flags `const`. I see this is also used in the flag parsing method

File: pkg/controller/atlasdatabaseuser/atlasdatabaseuser_controller.go [hunk 4]
```diff
@@ -78,155 +76,212 @@ func (r *AtlasDatabaseUserReconciler) Reconcile(context context.Context, req ctr
...
+	result = r.ensureDatabaseUser(workflowCtx, *project, *databaseUser)
+	if !result.IsOk() {
+		workflowCtx.SetConditionFromResult(status.DatabaseUserReadyType, result)
+
+		return result.ReconcileResult(), nil
 	}
 
-	return nil
-}
+	err = customresource.ManageFinalizer(ctx, r.Client, databaseUser, customresource.SetFinalizer)
```
Comment (igor-karpukhin) [inline pkg/controller/atlasdatabaseuser/atlasdatabaseuser_controller.go:173]: non-blocking: consider consistent naming: `EnsureFinalizer`

File: pkg/controller/atlasdatabaseuser/atlasdatabaseuser_controller.go [hunk 5]
```diff
@@ -78,155 +76,212 @@ func (r *AtlasDatabaseUserReconciler) Reconcile(context context.Context, req ctr
...
+	return true, workflow.OK()
+}
+
+func (r *AtlasDatabaseUserReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("AtlasDatabaseUser").
+		For(&mdbv1.AtlasDatabaseUser{}, builder.WithPredicates(r.GlobalPredicates...)).
+		Watches(&source.Kind{Type: &corev1.Secret{}}, watch.NewSecretHandler(r.WatchedResources)).
+		Complete(r)
+}
```
Comment (igor-karpukhin) [inline pkg/controller/atlasdatabaseuser/atlasdatabaseuser_controller.go:248]: Are you sure removing works fine in this case? I mean, won't it trigger the reconcile with result like "can not reconcile, the resource has been modified/deleted`?

**Review by igor-karpukhin — CHANGES_REQUESTED** [review]
Looks good, massive job here 👍 However, let's not merge it to "main", but into a separate branch ("deletion-protection", maybe), so we won't break "main".

File: pkg/controller/atlasdatabaseuser/atlasdatabaseuser_controller.go (hunk 4 above)
Comment (helderjs) [reply pkg/controller/atlasdatabaseuser/atlasdatabaseuser_controller.go:173]: As it's not a controller function but actually a helper, I think there's no inconsistency with naming

File: pkg/controller/atlasdatabaseuser/atlasdatabaseuser_controller.go (hunk 5 above)
Comment (helderjs) [reply pkg/controller/atlasdatabaseuser/atlasdatabaseuser_controller.go:248]: Yes, All the flows has worked as expected

File: pkg/controller/atlasdatabaseuser/atlasdatabaseuser_controller.go [hunk 6]
```diff
@@ -78,155 +76,212 @@ func (r *AtlasDatabaseUserReconciler) Reconcile(context context.Context, req ctr
...
-	return workflow.OK()
-}
 
-func (r *AtlasDatabaseUserReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	c, err := controller.New("AtlasDatabaseUser", mgr, controller.Options{Reconciler: r})
-	if err != nil {
-		return err
+	deletionRequest, result := r.handleDeletion(ctx, databaseUser, project, atlasClient, log)
+	if deletionRequest {
+		return result.ReconcileResult(), nil
```
Comment (josvazg) [inline pkg/controller/atlasdatabaseuser/atlasdatabaseuser_controller.go:154]: I am having difficulties following the logic of this reconcile function.
Why do we handle deletion here? I do not see where we are checking that a deletion is needed or requested.

File: pkg/controller/customresource/protection.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+func IsOwner(resource mdbv1.AtlasCustomResource, protectionFlag bool, operatorChecker OperatorChecker, atlasChecker AtlasChecker) (bool, error) {
+	if !protectionFlag {
+		return true, nil
+	}
+
+	wasManaged, err := operatorChecker(resource)
+	if err != nil {
+		return false, err
+	}
```
Comment (josvazg) [inline pkg/controller/customresource/protection.go:28]: If it was managed we do not need to go to atlas right? we can just return true here.

File: pkg/controller/atlasdatabaseuser/atlasdatabaseuser_controller.go (hunk 6 above)
Comment (helderjs) [reply pkg/controller/atlasdatabaseuser/atlasdatabaseuser_controller.go:154]: Here is where it's validated

https://github.com/mongodb/mongodb-atlas-kubernetes/pull/1003/files#diff-b936aa2d7b65922a3ed27d3debce3bf9cb8848177e170dbf3294d4157ee22c12R202-R204

File: pkg/controller/customresource/protection.go (hunk 7 above)
Comment (helderjs) [reply pkg/controller/customresource/protection.go:28]: Make sense

File: pkg/controller/atlasdatabaseuser/atlasdatabaseuser_controller.go (hunk 6 above)
Comment (josvazg) [reply pkg/controller/atlasdatabaseuser/atlasdatabaseuser_controller.go:154]: ok, I guess the confusion was due to the name of the method, `handleDeletion` makes you think that deletion is already decided.

**Review by igor-karpukhin — APPROVED** [review]

**Pull request merged**